src/ComposerConstraint.php

```diff
@@ -2,89 +2,148 @@
 
 namespace Drupal\lightning_dev;
 
-class ComposerConstraint {
+/**
+ * Class to perform operations on a composer constraint.
+ */
+final class ComposerConstraint {
 
   /**
+   * Raw constraint.
+   *
    * E.g. '^2.8 || ^3.0'.
    *
    * @var string
    */
   private $constraint = '';
 
   /**
+   * ComposerConstraint constructor.
+   *
    * @param string $constraint
+   *   Raw constraint.
    */
   public function __construct($constraint) {
     $this->constraint = $constraint;
   }
 
   /**
+   * Returns the constraint's ranges as an array.
+   *
+   * @see https://getcomposer.org/doc/articles/versions.md#version-range
+   *
    * @return string[]
-   *   E.g. ['^2.8', '^3.0'].
+   *   Constraint's ranges. For example, if the constraint is
+   *   '^2.8 || ^3.0', it will be ['^2.8', '^3.0'].
    */
-  public function getRanges() {
+  private function getRanges() {
     preg_match_all('/[0-9a-zA-Z\~\>\=\-\<\.\^\*]+/', $this->constraint, $matches);
 
     return $matches[0];
   }
 
   /**
+   * Returns the core dev version of the constraint.
+   *
+   * In the core dev version the last digits of the constraint's ranges are
+   * replaced by the string 'x-dev', and their operators are removed.
+   *
    * @return string
-   *   E.g. '8.4.x-dev || 8.5.x-dev'.
+   *   Core dev version. For example, if the constraint is
+   *   '8.4.3 || ^8.5.3', it will be '8.4.x-dev || 8.5.x-dev'.
    */
   public function getCoreDev() {
     return $this->getDev([static::class, 'coreRangeToDev']);
   }
 
   /**
+   * Returns the lightning dev version of the constraint.
+   *
+   * In the lightning dev version the first digits of the constraint's ranges are
+   * concatenated with the string 'x-dev', and their operators are removed.
+   *
    * @return string
-   *   E.g. '2.x-dev || 3.x-dev'.
+   *   Lightning dev version. For example, if the constraint is
+   *   '^1.3.0 || ^2.3.0', it will be '1.x-dev || 2.x-dev'.
    */
   public function getLightningDev() {
     return $this->getDev([static::class, 'lightningRangeToDev']);
   }
 
   /**
+   * Returns the dev version of the constraint.
+   *
+   * In the dev version the constraint's ranges are replaced with the
+   * result of the callback.
+   *
    * @param callable $callback
+   *   Transformation to apply to each range.
    *
    * @return string
+   *   Dev version.
    */
-  private function getDev($callback) {
-    $dev = $this->constraint;
+  private function getDev(callable $callback) {
     $ranges = $this->getRanges();
+    $replace_pairs = [];
 
     foreach ($ranges as $oldRange) {
       $newRange = $callback($oldRange);
```

Comment:
Nitpick: Let's use consistent snake_case casing for all variable names. So $newRange should be $new_range.